pkg/promql/engine.go

```diff
@@ -780,14 +780,13 @@ func (ng *Engine) getTimeRangesForSelector(s *parser.EvalStmt, n *parser.VectorS
 	return start, end
 }
 
-func (ng *Engine) populateSeries(querier Querier, s *parser.EvalStmt) parser.Node {
+func (ng *Engine) populateSeries(querier Querier, s *parser.EvalStmt) map[parser.Node]bool {
 	var (
 		// Whenever a MatrixSelector is evaluated, evalRange is set to the corresponding range.
 		// The evaluation of the VectorSelector inside then evaluates the given range and unsets
 		// the variable.
-		set       storage.SeriesSet
 		evalRange time.Duration
-		topNode   parser.Node
+		topNodes  map[parser.Node]bool = make(map[parser.Node]bool)
```

Comment:
In maps like these, where you just check for existence, its idiomatic to use an empty struct as the value since it doesn't use any memory to store and check for existence of the key without caring for the value (it would be set to `true` always).

Its a small thing TBH but posting for knowledge sharing.
https://levelup.gitconnected.com/memory-allocation-and-performance-in-golang-maps-b267b5ad9217